packages/database/src/core/WriteTree.ts

```diff
@@ -59,575 +59,745 @@ export interface WriteRecord {
+      );
     } else {
-      for (const childName in writeRecord.children) {
-        if (
-          writeRecord.children.hasOwnProperty(childName) &&
-          pathContains(pathChild(writeRecord.path, childName), path)
-        ) {
-          return true;
-        }
-      }
-      return false;
+      return null;
     }
   }
+}
 
-  /**
-   * Re-layer the writes and merges into a tree so we can efficiently calculate event snapshots
-   */
-  private resetTree_() {
-    this.visibleWrites_ = WriteTree.layerTree_(
-      this.allWrites_,
-      WriteTree.DefaultFilter_,
-      newEmptyPath()
-    );
-    if (this.allWrites_.length > 0) {
-      this.lastWriteId_ = this.allWrites_[this.allWrites_.length - 1].writeId;
-    } else {
-      this.lastWriteId_ = -1;
+/**
+ * Returns a node if there is a complete overwrite for this path. More specifically, if there is a write at
+ * a higher path, this will return the child of that write relative to the write and this path.
+ * Returns null if there is no write at this path.
+ */
+export function writeTreeShadowingWrite(
+  writeTree: WriteTree,
+  path: Path
+): Node | null {
+  return compoundWriteGetCompleteNode(writeTree.visibleWrites_, path);
+}
+
+/**
+ * This method is used when processing child remove events on a query. If we can, we pull in children that were outside
+ * the window, but may now be in the window.
+ */
+export function writeTreeCalcIndexedSlice(
+  writeTree: WriteTree,
+  treePath: Path,
+  completeServerData: Node | null,
+  startPost: NamedNode,
+  count: number,
+  reverse: boolean,
+  index: Index
+): NamedNode[] {
+  let toIterate: Node;
+  const merge = compoundWriteChildCompoundWrite(
+    writeTree.visibleWrites_,
+    treePath
+  );
+  const shadowingNode = compoundWriteGetCompleteNode(merge, newEmptyPath());
+  if (shadowingNode != null) {
+    toIterate = shadowingNode;
+  } else if (completeServerData != null) {
+    toIterate = compoundWriteApply(merge, completeServerData);
+  } else {
+    // no children to iterate on
+    return [];
+  }
+  toIterate = toIterate.withIndex(index);
+  if (!toIterate.isEmpty() && !toIterate.isLeafNode()) {
+    const nodes = [];
+    const cmp = index.getCompare();
+    const iter = reverse
+      ? (toIterate as ChildrenNode).getReverseIteratorFrom(startPost, index)
+      : (toIterate as ChildrenNode).getIteratorFrom(startPost, index);
+    let next = iter.getNext();
+    while (next && nodes.length < count) {
+      if (cmp(next, startPost) !== 0) {
+        nodes.push(next);
+      }
+      next = iter.getNext();
     }
+    return nodes;
+  } else {
+    return [];
   }
+}
 
+/**
+ * WriteTree tracks all pending user-initiated writes and has methods to calculate the result of merging them
+ * with underlying server data (to create "event cache" data).  Pending writes are added with addOverwrite()
+ * and addMerge(), and removed with removeWrite().
+ */
+export class WriteTree {
```

Comment:
This could also be an interface. It is only created in one place (but outside this module - so this change would leak some internals unless you create a `newWriteTree()` function).

Reply:
Changed to an interface.